File: frontend/src/api/academyScope.js
/**
 * academyScope — portée d'académie courante du client HTTP.
 *
 * PROBLÈME RÉSOLU (P0)
 * --------------------
 * Au changement d'académie, le sélecteur changeait de libellé instantanément
 * mais les tableaux restaient plusieurs secondes sur les données de
 * l'académie quittée. La cause n'était pas seulement le cache : les requêtes
 * DÉJÀ EN VOL se terminaient après la bascule et réécrivaient le cache avec
 * la réponse de l'ancienne académie. Vider le cache ne servait à rien
 * puisqu'il était immédiatement re-rempli par ces réponses tardives.
 *
 * Ce module apporte deux garanties :
 *
 *   1. ANNULATION — chaque requête sortante reçoit un AbortController
 *      enregistré ici. `setAcademyScope()` avorte tout ce qui est en vol :
 *      une requête émise pour FEBA ne peut plus aboutir après un passage
 *      sur FEBA FHA.
 *
 *   2. VÉRIFICATION — chaque requête annonce sa portée (`X-Academy-Scope`)
 *      et le serveur répond avec la portée réellement utilisée. Si les deux
 *      diffèrent, la réponse est rejetée : elle décrit une académie qui
 *      n'est plus affichée. C'est le filet de sécurité pour les cas que
 *      l'annulation ne couvre pas (réponse déjà arrivée dans le tampon
 *      réseau, requête émise hors de notre intercepteur…).
 *
 * Le module est volontairement sans dépendance React : il est utilisé par
 * les intercepteurs axios, qui s'exécutent hors du cycle de rendu.
 */

export const SCOPE_ALL = "ALL";
export const SCOPE_UNKNOWN = "UNKNOWN";
export const HEADER = "X-Academy-Scope";

/** Portée annoncée sur les requêtes sortantes. */
let currentScope = SCOPE_UNKNOWN;

/** Contrôleurs des requêtes en vol, avortés à chaque bascule. */
const inflight = new Set();

/** Erreur marquant une réponse écartée parce qu'elle n'est plus d'actualité. */
export class StaleAcademyResponse extends Error {
  constructor(expected, received, url) {
    super(
      `Réponse ignorée : calculée pour l'académie « ${received} », ` +
        `l'interface est sur « ${expected} » (${url || "requête inconnue"}).`,
    );
    this.name = "StaleAcademyResponse";
    this.expected = expected;
    this.received = received;
    /* React Query ne doit ni réessayer ni afficher cette erreur : le
       composant concerné est de toute façon en cours de remontage. */
    this.isStaleAcademy = true;
  }
}

export function getAcademyScope() {
  return currentScope;
}

/**
 * Endpoints exemptés du contrôle de portée.
 *
 * La requête de BASCULE est celle qui CHANGE la portée : le serveur la
 * traite, puis annonce dans sa réponse la nouvelle académie — forcément
 * différente de celle que le client avait au moment de l'envoi. La traiter
 * comme une réponse périmée revenait à rejeter la bascule elle-même : le
 * serveur enregistrait bien le changement, mais le frontend n'en était
 * jamais informé et le sélecteur restait figé sur l'académie précédente.
 *
 * La lecture du contexte est exemptée pour la même raison : c'est elle qui
 * ÉTABLIT la portée, elle ne peut pas être vérifiée contre elle-même.
 */
const SCOPE_EXEMPT = ["/auth/entity-context/"];

export function isScopeExempt(url) {
  if (!url) return false;
  return SCOPE_EXEMPT.some((path) => url.includes(path));
}

/**
 * Déclare la portée active et invalide tout ce qui appartenait à la
 * précédente. Appelée par AcademyProvider dès que le SERVEUR a confirmé
 * la bascule — jamais sur une simple intention de l'utilisateur.
 *
 * @returns {boolean} true si la portée a réellement changé.
 */
export function setAcademyScope(next) {
  const value = next || SCOPE_UNKNOWN;
  if (value === currentScope) return false;
  currentScope = value;
  abortInflightRequests();
  return true;
}

/** Avorte toutes les requêtes en vol (bascule d'académie, déconnexion…). */
export function abortInflightRequests() {
  for (const controller of inflight) {
    try {
      controller.abort();
    } catch {
      /* Un contrôleur déjà consommé lève sur certains navigateurs :
         sans importance, la requête est terminée. */
    }
  }
  inflight.clear();
}

export function trackRequest() {
  const controller = new AbortController();
  inflight.add(controller);
  return controller;
}

export function releaseRequest(controller) {
  if (controller) inflight.delete(controller);
}

/** Nombre de requêtes en vol — utilisé par les tests. */
export function inflightCount() {
  return inflight.size;
}

/**
 * True si la réponse doit être écartée.
 *
 * `UNKNOWN` des deux côtés est toléré : au tout premier rendu, le client
 * n'a pas encore reçu son contexte d'académie. Rejeter ces réponses
 * bloquerait l'application au démarrage.
 */
export function isStaleResponse(responseScope, url) {
  if (!responseScope) return false;
  if (isScopeExempt(url)) return false;
  if (currentScope === SCOPE_UNKNOWN) return false;
  return responseScope !== currentScope;
}

/** Remise à zéro — tests unitaires uniquement. */
export function resetAcademyScope() {
  currentScope = SCOPE_UNKNOWN;
  inflight.clear();
}

File: backend/apps/core/academy_scope.py
"""
apps/core/academy_scope.py — Portée d'académie transportée par en-tête HTTP

PROBLÈME RÉSOLU (P0)
--------------------
Au changement d'académie, le libellé du sélecteur changeait tout de suite
mais les tableaux continuaient d'afficher les données de l'académie
quittée pendant plusieurs secondes. Deux causes distinctes :

1. les requêtes déjà EN VOL au moment de la bascule se terminaient après
   celle-ci et réécrivaient le cache avec des données périmées ;
2. rien ne permettait de savoir, en regardant une réponse, à quelle
   académie elle correspondait — donc rien ne permettait de la rejeter.

Ce module traite le point 2 côté serveur : **chaque réponse API annonce
la portée qui a réellement servi à la calculer**, dans l'en-tête
`X-Academy-Scope`. Le frontend compare cette valeur à sa portée courante
et jette toute réponse qui ne correspond pas.

SÉCURITÉ — CE QUE CET EN-TÊTE N'EST PAS
---------------------------------------
`X-Academy-Scope` envoyé par le NAVIGATEUR n'a AUCUNE autorité : il n'est
jamais utilisé pour choisir les données renvoyées. La portée effective
reste déterminée exclusivement par `get_request_school()` (rattachement
en base, ou entité active persistée du superadmin). L'en-tête client sert
uniquement à journaliser les désynchronisations. Le faire mentir ne donne
accès à rien.

VALEURS
-------
    ``ALL``       superadmin en mode « Toutes les Académies »
    ``FEBA`` …    code interne stable de l'académie résolue
    ``NONE``      requête anonyme, ou compte sans rattachement
"""
import logging

logger = logging.getLogger("apps")

HEADER = "X-Academy-Scope"
CLIENT_HEADER = "HTTP_X_ACADEMY_SCOPE"

SCOPE_ALL = "ALL"
SCOPE_NONE = "NONE"


def resolve_scope_value(request):
    """
    Portée effective de la requête, sous forme de chaîne comparable.

    Volontairement tolérant : cette fonction s'exécute pour TOUTES les
    réponses, y compris les erreurs 500. Elle ne doit jamais être la
    cause d'un échec — d'où le repli sur ``NONE``.
    """
    user = getattr(request, "user", None)
    if user is None or not getattr(user, "is_authenticated", False):
        return SCOPE_NONE

    try:
        from apps.core.tenancy import get_request_school

        school = get_request_school(request)
    except Exception:  # pragma: no cover — filet de sécurité
        return SCOPE_NONE

    if school is None:
        # Un superadmin sans entité active est en mode consolidé ; un
        # utilisateur normal sans école est un compte mal configuré.
        return SCOPE_ALL if getattr(user, "is_superadmin", lambda: False)() else SCOPE_NONE
    return school.code or f"id:{school.pk}"


class AcademyScopeMiddleware:
    """
    Ajoute `X-Academy-Scope` à chaque réponse.

    Le middleware calcule la portée APRÈS la vue : l'authentification JWT
    de DRF n'a lieu qu'à ce moment-là (DRF recopie l'utilisateur
    authentifié sur la requête Django sous-jacente). La calculer avant
    donnerait systématiquement ``NONE``.
    """

    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        response = self.get_response(request)

        scope = resolve_scope_value(request)
        response[HEADER] = scope

        announced = request.META.get(CLIENT_HEADER)
        if announced and announced != scope and not _is_switch_endpoint(request):
            # Désynchronisation : le navigateur croyait encore être sur une
            # autre académie. La réponse reste correcte (elle a été calculée
            # avec la portée serveur) ; c'est au frontend de la jeter.
            logger.info(
                "Portée d'académie désynchronisée sur %s : le client annonçait "
                "« %s », le serveur a servi « %s ».",
                request.path, announced, scope,
            )

        return response


def _is_switch_endpoint(request):
    """
    La requête de bascule elle-même change la portée en cours de route :
    le client annonce forcément l'ancienne valeur. Ce n'est pas une
    anomalie, on ne la journalise pas.
    """
    return request.path.endswith("/entity-context/switch/")
